runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -831,6 +831,161 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 		return getSpineSize(J9GC_J9OBJECT_CLAZZ(arrayPtr, this), layout, getSizeInElements(arrayPtr));
 	}
 
+#if defined(J9VM_ENV_DATA64)
+
+	/**
+	 * Gets data pointer of a contiguous indexable object.
+	 * Helper to get dataAddr field of contiguous indexable objects.
+	 *
+	 * @return Pointer which points to indexable object data
+	 */
+	MMINLINE U_64*
+	dataAddrSlotForContiguous(J9IndexableObject *arrayPtr)
+	{
+		AssertContiguousArrayletLayout(arrayPtr);
+		bool const compressed = compressObjectReferences();
+		U_64 *address = NULL;
+		if (compressed) {
+			address = &((J9IndexableObjectContiguousCompressed *)arrayPtr)->dataAddr;
+		} else {
+			address = &((J9IndexableObjectContiguousFull *)arrayPtr)->dataAddr;
+		}
+		return address;
+	}
+
+	/**
+	 * Gets data pointer of a discontiguous indexable object.
+	 * Helper to get dataAddr field of discontiguous indexable objects.
+	 *
+	 * @return Pointer which points to discontiguous indexable object data
+	 */
+	MMINLINE U_64*
+	dataAddrSlotForDiscontiguous(J9IndexableObject *arrayPtr)
+	{
+		/* If double mapping is enabled only, arraylet will have a discontiguous layout.
```

Comment:
This this use `AssertDiscontiguousArrayletLayout`?

Reply:
For now we only use `AssertDiscontiguousArrayletLayout` in `getDataAddrForDiscontiguous`, because `setDataAddrForDiscontiguous` will also be setting `dataAddr` field for hybrid arraylets and `AssertDiscontiguousArrayletLayout` only asserts if they're strictly Discontiguous. The reason for that is that we made the assumption that `getDataAddrForDiscontiguous(J9IndexableObject *arrayPtr)` would only be used in case of pure disncotiguous arraylets (double mapping case or future off-heap technology)